### ISSUE TITLE
Surface container failures in a more obvious manner

### ISSUE THREAD

**Issue by vmarmol** [GoogleCloudPlatform/kubernetes#6014]
Following on #5927, should pods with containers that crashed have that status be reflected in the pod status? Primary usecase is crashloops or frequent crashes. The pods today are reported as "running" although the individual ContainerStatus will report the container crashing. This makes it hard for users to identify these types of problems.

We can consider setting the pod status to Pending (although the description today would have to be changed slightly), or a new state type. The status would be set to Running in the Kubelet's next sync loop (modulo details of this taking 10s, we can make that shorter for these cases).

**Comment by vmarmol** [comment]
We just had a second issue filed for this (#5668), may be worth making a v1.0 item?

**Comment by yujuhong** [comment]
I just saw an old, related issue about this #4121.

**Comment by yujuhong** [comment]
We could either re-purpose an existing phase or create a new one. 
 * Pending: I associate this state with "time-in-between" different tasks, which does not convey a strong signal of failing.
 * Unknown: Unknown suggests an infrastructure issue (e.g. kubelet is temporarily unresponsive).
 * Failed: Failed could be the containers terminated in a failure, or kubelet was unable to start the containers (e.g. host port conflicts, etc). It is an indication that "something" went wrong.

Given the reasons, I think re-using "Failed" makes the most sense. The only downside, like I said in #5927, is that kubelet won't be respecting user's wish to always restart the containers. The alternative is, as @bgrant0607 suggested in #4121, to back off on restarting containers. This would probably result in flipping status of "Failed" and "Running". 

In short, the options are:
  1. Detect the crash loop, and set the pod status to failed.
  2. Back off on each crash; let the status flip between "Running" and "Failed"
  3. Combine both (1) and (2). Back off on each restart and eventually give up and set "Failed".

p.s. surfacing this to the replication controller had been discussed in a separate issue (#2529).
 
/cc @bgrant0607, @dchen1107 for input.

**Comment by vmarmol** [comment]
I think "Failed" is a final state. I'd say to flip between Pending and Running during our allowed time. When we determine it is a crash-loop we can change to the final Failed state as you mentioned and reject the pod.

**Comment by dchen1107** [comment]
I saw #5927 yesterday, but failed to follow up due to meetings.

I disagree with the proposal made here, and I don't want another phase introduced here. We discussed the lifecycle of Pod which is documented at kubernetes/docs/pod-states.md. 

For particular issue reported #5927, we should addressed through #4919 by reporting last terminated container status. I just introduced API for that, and a pending PR to report previous failures. 

Also #4121 (backoff restart crashing container) should be handled too.

**Comment by yujuhong** [comment]
@dchen1107 , I think part of the reason why users complain (e.g. #5927) is that it's hard to "notice" that the pod is crashing. Personally, I would just run `kubectl get pods` and examine the current pods. When I see "Running", I'd expect the pod is working as expected. The container crashes are not visible to the users at a higher level.

The other reason to propagate the status is to notify replication controller of the incident. Do you plan to instruct the replication controller to examine the containerStatus?

When you said "#4121 should be handled too", do you mean that your PR will handle that, or someone should work on it?

**Comment by dchen1107** [comment]
I think we should correct restart count for ContainerStatus. Now it is capped due to container GC, but it could be fixed at the best effort now since kubelet POST status now. :-)

I think replication controller should examine the ContainerStatus, there is an issue filed against that. @bgrant0607 

I hope someone work on #4121.

**Comment by bgrant0607** [comment]
No new pod phase, and pod phase should progress monotonically.

Fixing the restart count seems like a useful thing to do. Use #4121 for that or file a new issue if there isn't already one specific for that. Or just submit a PR.

Surfacing container status in kubectl output also would be worthwhile. Additionally, we could customize the output (esp. describe output) based on the phase, reason, whether host is set, status, conditions, etc. to be more explanatory/informative. For instance, if the phase is pending, we could display the reason. If the phase is running, we could just show container statuses. If container status is failed, we could show the termination status. If the container recently failed, we could report running but failed N seconds ago with the last termination reason, message, etc.

Making replication controller smarter is #2529.

**Comment by dchen1107** [comment]
cc/ @rrati

**Comment by rrati** [comment]
@bgrant0607 +1 for more information in the events/kubectl.  There's currently no visibility into what is going on and why a pod is in a given state.

Is there agreement that the state in master needs to reflect the true state of the pods on a minion for 1.0?  This would include some level of handling recurring container crashes and/or container failures such that the pod status in master is not "Running"

**Comment by dchen1107** [comment]
@rrati I filed #6076 to show containers information through kubectl describe pods id. 
Also ContainerStatus is going to include Last termination (#4919), and restart_count will be corrected at the best effort. 

Kubectl describe pods id also includes related events of a Pod last two hours (TTL). Here is a snippet from one pod with crashloop with PR #5994:

```
Name:				valid-pod
Image(s):			kubernetes/serve_hostname
Host:				kubernetes-minion-p1yo.c.golden-system-455.internal/
Labels:				name=valid-pod
Status:				Pending
Replication Controllers:	<none>
Conditions:
  Type		Status
  Ready 	False 
Events:
  FirstSeen				LastSeen			Count	From								SubobjectPath					Reason	Message
  Thu, 26 Mar 2015 22:44:20 -0700	Fri, 27 Mar 2015 11:38:38 -0700	4645	{kubelet kubernetes-minion-p1yo.c.golden-system-455.internal}	spec.containers{kubernetes-serve-hostname}	failed	Failed to create docker container with error: API error (500): Minimum memory limit allowed is 4MB
```

I think with all above included, the state in master is enough to reflect the true state of the pod on a minion for 1.0, disagree?

**Comment by rrati** [comment]
@dchen1107 what will 'kubectl get pods' show in these instances?  Will that display Running or Pending?  If the pod starts running and then a container crashes, will the state transition back to Pending?

**Comment by bgrant0607** [comment]
@rrati Pods are comprised of multiple containers, even when the user specifies just one container. If what you want is the status of your container, you need to look at ContainerStatus, not just PodPhase. More information is available than is currently displayed, and we also plan to report more information. The most important information should be displayed by kubectl, and less useful information should be elided.

**Comment by rrati** [comment]
@bgrant0607 If a container isn't running then imo the pod phase should not be running.  If I'm looking at pod states through kubectl, the most important information to me is that the container states are properly reflected and that if a container in a pod isn't running there is an indication that is so.  That can be accomplished a number of ways, but currently we are giving the "all clear" status with no indication a user should look deeper and see that containers aren't running as expected.

**Comment by dchen1107** [comment]
We are going to include Container information to kubect get pods <id>

**Comment by smarterclayton** [comment]
This is becoming very important for us - knowing that a restart always pod is crash-looping is turning into a big blocker.  I think we should be showing the time a container last started in the pod list, probably (https://github.com/GoogleCloudPlatform/kubernetes/pull/6361 is good but Get is important as well).

**Comment by smarterclayton** [comment]
May want to tie in ContainerStateTerminated along with most recent container start time - if we *died* vs were intentionally restarted, that's important.

**Comment by yujuhong** [comment]
`kubectl get pods` already displays the container names and images as below.

```
CONTAINER(S)            IMAGE(S)
con1, con2, con3        image1, image2, image3
```

How about we add the container statuses using the same format?
```
CONTAINER_STATUS            CONTAINER_RESTART_COUNT     CONTAINER_CREATED
Running, Waiting, Running   1, 4, 3                     20 minutes, n/a, 10 minutes
```

The restart counts accumulate over time; the container_created time can help identify problematic containers along with the restart count. The user can always run `kubectl describe pod_name` for more details after seeing unusual signals from the output. 

I could add more columns but I am afraid the wide output would worsen the readability (or is that not a concern?). If we truly want more information, maybe we'd need a `-v` flag.

Thoughts and suggestions?

**Comment by smarterclayton** [comment]
We could possibly just show a column which is "LAST START" or something which is the container's "<time since last start>, <reason>".  I don't think showing all the containers is most useful - you probably want the "this is failing and restarting" to pop out at you.

----- Original Message -----

**Comment by dchen1107** [comment]
We already reported lastTerminatedState for each container, populating such information for "kubectl get pods id" maybe is good enough for now?

**Comment by yujuhong** [comment]
I spoke to @vmarmol, and he corrected me on the current status. We now display the container row by row in `get pod`, so the containers could just reuse the columns from the pods.

E.g.
```
POD         IP           CONTAINER(S)  IMAGE(S)  HOST  LABELS STATUS   CREATED
kube-dns-1  10.245.1.3   etcd          image1    ...   ...    Running  2 days
                         kube2sky      image2                                                                                                                                                                                                    
                         skydns        image3                                                                                                                                                                                           
```

We could simply populate the  container status for each row of containers.

**Comment by smarterclayton** [comment]
I'm thinking that get is for "at a glance" - I really only need to know "when was this container last restarted, and why did it restart" in a succinct way (as you note, we can dig into describe).  If we spend more than 20 characters on it, it's probably too much.

----- Original Message -----

**Comment by dchen1107** [comment]
@smarterclayton +1 on get is for "at a glance". A clarification from previous comment: "when and why this container restarted" can be retrieved from lastTerminatedState for each container.

**Comment by yujuhong** [comment]
I agree with @smarterclayton and @dchen1107 that this information should be for "at a glance". My point in the previous comment was that we *already* displayed one container per row in `get pods`, so we could just reuse the columns there. We will need to shift the rows down one row, so that the first row is dedicated to the pod itself. For other rows, we could show:
```
STATUS:  display the containers status
CREATED: display the last start time
```

E.g.
```
POD         IP           CONTAINER(S)  IMAGE(S)  HOST       LABELS STATUS   CREATED
kube-dns-1  10.245.1.3                           localhost  ...    Running  2 days
                         etcd          image1                      Running  1 days
                         kube2sky      image2                      Waiting  2 days                                                                                                                                                                                      
                         skydns        image3                      Running  2 days                                                                                                                                                                           
```
We could also add an additional `REASON` column for the last terminated reason.

The downside is that each pod would occupy at least two rows (one for the pod itself, and the other for the container). The upside is we already use the real estate to display the one container per row, so this is not a big change and we can reuse columns.

**Comment by smarterclayton** [comment]
SGTM

----- Original Message -----

**Comment by yujuhong** [comment]
#7116 has been merged. It adds the container info in `kubectl get pods`

**Comment by dchen1107** [comment]
Can we close this issue since #7116 is in already?

**Comment by timothysc** [comment]
@dchen1107 I'm ok for now, but later on we'll likely need history/accounting.

**Comment by yujuhong** [comment]
@timothysc, could you clarify what kind of history/accounting that you'd need?

**Comment by timothysc** [comment]
@yujuhong At some point in the future ... doing an audit history of pod runtime will likely be need:

ran W, on host X, for Y secs, exited code (Z) 
ran W, on host XX, for YY secs, exist code (ZZ)
... 
per pod. 

Usually this type of thing is offloaded into a history server of some kind.  

History servers can then do all kinds of cool things, like mine run information for correct sizing etc.  i.e. - Find bad users, bad jobs, bad...

**Comment by dchen1107** [comment]
@timothysc  in a long run, we should have infrastore-like system to store those history, so that the user can access those information offline. But I don't think kubectl get pods should manage such information.

**Comment by timothysc** [comment]
@dchen1107 agreed, I'm just chiming in on the closing of *this ticket.

**Comment by dchen1107** [comment]
@timothysc thanks for chiming in. Also today you can have some history through events. Not perfect yet.
